extensions/reviewed/InAppPurchase.json

```diff
@@ -140,90 +153,106 @@
-      "fullName": "Update a variable when an item is updated",
+      "description": "When an event is triggered for an item (approved or finished), this sets a scene variable to true. \nYou can then compare the value of the variable in a condition, and have actions launched to react to the changes.",
+      "fullName": "Update a variable when an item event is triggered",
       "functionType": "Action",
       "name": "WatchItemEvent",
-      "private": false,
       "sentence": "Watch the event _PARAM3_ for item _PARAM1_ and set scene variable named _PARAM2_ to true when it happens",
       "events": [
         {
-          "disabled": false,
-          "folded": false,
           "type": "BuiltinCommonInstructions::JsCode",
-          "inlineCode": "if(!window.store) return;\n\nstore.when(eventsFunctionContext.getArgument(\"id\"), eventsFunctionContext.getArgument(\"event\"), () => {\n    const variable = runtimeScene.getVariables().get(eventsFunctionContext.getArgument(\"variableName\"));\n    variable.setBoolean(true);\n});\n",
+          "inlineCode": [
+            "if(!window.CdvPurchase || !window.CdvPurchase.store) return;",
+            "const store = window.CdvPurchase.store;",
+            "",
+            "const eventName = eventsFunctionContext.getArgument(\"event\");",
+            "const productId = eventsFunctionContext.getArgument(\"id\");",
+            "",
+            "const transactionCallback = (transaction) => {",
+            "    const products = transaction.products;",
+            "    // If the product being watched is in the list of products in the transaction, save into a variable.",
+            "    if (products.find(transactionProduct => transactionProduct.id === productId)) {",
+            "        const variable = runtimeScene.getVariables().get(eventsFunctionContext.getArgument(\"variableName\"));",
+            "        variable.setBoolean(true);",
+            "    }",
+            "}",
+            "",
+            "store.when().approved((transaction) => {",
+            "    // If we're watching this event, call the callback.",
+            "    if (eventName === 'approved') {",
+            "        transactionCallback(transaction);",
+            "    }",
+            "    // Automatically verify the transaction on approved.",
+            "    transaction.verify();",
+            "})",
+            "",
+            "// Automatically finish a receipt when verified.",
+            "store.when().verified(receipt => receipt.finish())",
+            "",
+            "store.when().finished((transaction) => {",
+            "    // If we're watching this event, call the callback.",
+            "    if (eventName === 'finished') {",
+            "        transactionCallback(transaction);",
+            "    }",
+            "    // Nothing left to do when finished.",
+            "})",
```

Comment:
I have received multiple reports of the extension being broken & orders going unfulfilled after this update.

Taking a look now, this looks extremely wrong to me. 
1. This adds a global event handlers for every event type every time this action is called? This is completely unnecessary, harmful to memory and performance, and will almost certainly lead to multiple calls of `.verify()`/`.finish()`, which is not valid and likely will cause unintended behavior!

2. This forces a forbidden behavior upon users: once this action has been called, and the purchase of the product is started, it will automatically do the steps to finalize the purchase. However, it is absolutely not how it is supposed to be done! Not letting the creator handle verification themselves is arguably not that big of a deal, but we cannot mark the purchases as finalized automatically. A purchase may only ever be marked as finished once the game code has successfully delivered the product that was bought, at this point we know for sure the item has not been delivered yet. Heck, this code will mark as finished *ANY* product that was bought, even ones for which the creator has not even put up this event listener for!
3. It is a very error prone implementation. If this action does not get called, a purchase of a product will never be processed since those event handlers will never have been set. The action is inconspicuous, doesn't seem and doesn't say in any way that it is crucial for the extension to function!

Reply:
Thanks for the feedback @arthuro555 
I agree the implementation looks incomplete and was done in haste, to prevent creators from being stuck in publishing their app on the stores because of the Google Play deadline.
The tests I had done were fine on the example, but it may have been not enough.
I feel bad about having published this update now.

As you know well how this works, would you have time to improve it and make the extension less error-prone?